File: rust/crates/openjarvis-security/src/permissions.rs
//! Minimal Rust permission policy for tool execution.
//!
//! This mirrors the Python `PermissionMiddleware` concepts without adding a
//! cross-language runtime dependency. Keep labels and path rules aligned with
//! `src/openjarvis/security/permissions.py` and `file_policy.py`.

use crate::file_policy::{has_path_traversal, is_protected_path, is_sensitive_path};
use once_cell::sync::Lazy;
use regex::Regex;
use serde::{Deserialize, Serialize};
use serde_json::{json, Value};
use std::collections::HashSet;
use std::fs::OpenOptions;
use std::io::Write;
use std::path::{Path, PathBuf};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum PermissionLevel {
    ReadOnly,
    SafeAction,
    ConfirmedExecution,
    Dangerous,
}

impl PermissionLevel {
    pub fn as_str(&self) -> &'static str {
        match self {
            PermissionLevel::ReadOnly => "READ_ONLY",
            PermissionLevel::SafeAction => "SAFE_ACTION",
            PermissionLevel::ConfirmedExecution => "CONFIRMED_EXECUTION",
            PermissionLevel::Dangerous => "DANGEROUS",
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum PermissionAction {
    Allow,
    RequireConfirmation,
    Deny,
}

impl PermissionAction {
    pub fn as_str(&self) -> &'static str {
        match self {
            PermissionAction::Allow => "allow",
            PermissionAction::RequireConfirmation => "require_confirmation",
            PermissionAction::Deny => "deny",
        }
    }
}

#[derive(Debug, Clone)]
pub struct PermissionRequest<'a> {
    pub tool_name: &'a str,
    pub arguments: &'a Value,
    pub agent_id: Option<&'a str>,
    pub command: Option<&'a str>,
    pub source: &'a str,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct PermissionDecision {
    pub action: PermissionAction,
    pub level: PermissionLevel,
    pub reason: String,
    pub matched_pattern: Option<String>,
    pub dry_run: bool,
    pub would_action: Option<PermissionAction>,
}

impl PermissionDecision {
    pub fn allowed(&self) -> bool {
        self.action == PermissionAction::Allow
    }

    pub fn denied(&self) -> bool {
        self.action == PermissionAction::Deny
    }

    pub fn requires_confirmation(&self) -> bool {
        self.action == PermissionAction::RequireConfirmation
    }
}

#[derive(Debug, Clone)]
pub struct PermissionPolicy {
    audit_log_path: PathBuf,
}

impl Default for PermissionPolicy {
    fn default() -> Self {
        Self {
            audit_log_path: default_audit_log_path(),
        }
    }
}

impl PermissionPolicy {
    pub fn new(audit_log_path: impl Into<PathBuf>) -> Self {
        Self {
            audit_log_path: audit_log_path.into(),
        }
    }

    pub fn audit_log_path(&self) -> &Path {
        &self.audit_log_path
    }

    pub fn check(&self, request: &PermissionRequest<'_>) -> PermissionDecision {
        let (level, mut reason, matched_pattern) = self.classify(request);
        let mut action = action_for_level(level);
        let dry_run = request
            .arguments
            .get("dry_run")
            .and_then(Value::as_bool)
            .unwrap_or(false);
        let mut would_action = None;

        if dry_run {
            reason = format!("dry_run: would {} ({})", action.as_str(), reason);
            would_action = Some(action);
            action = PermissionAction::Allow;
        }

        let decision = PermissionDecision {
            action,
            level,
            reason,
            matched_pattern,
            dry_run,
            would_action,
        };
        self.audit(request, &decision);
        decision
    }

    pub fn classify(
        &self,
        request: &PermissionRequest<'_>,
    ) -> (PermissionLevel, String, Option<String>) {
        let tool_name = request.tool_name;
        let command = request
            .command
            .map(ToOwned::to_owned)
            .unwrap_or_else(|| extract_command(request.arguments));

        if tool_name == "shell_exec" || looks_like_shell_tool(tool_name) {
            let (level, reason, matched) = classify_shell_command(&command);
            if tool_name != "shell_exec" && level == PermissionLevel::ConfirmedExecution {
                return (
                    level,
                    format!("MCP shell-style tool requires confirmation: {tool_name}"),
                    matched,
                );
            }
            return (level, reason, matched);
        }

        if tool_name == "file_read" {
            if let Some(path) = string_arg(request.arguments, "path") {
                let path = Path::new(path);
                if is_sensitive_path(path) {
                    return (
                        PermissionLevel::Dangerous,
                        "sensitive file read blocked".into(),
                        Some("sensitive-path".into()),
                    );
                }
            }
        }

        if matches!(tool_name, "file_write" | "apply_patch") {
            if let Some(reason) = blocked_file_write_reason(request.arguments) {
                return (
                    PermissionLevel::Dangerous,
                    reason,
                    Some("protected-path".into()),
                );
            }
        }

        if tool_name == "http_request" {
            let method = string_arg(request.arguments, "method")
                .unwrap_or("GET")
                .to_uppercase();
            if matches!(method.as_str(), "POST" | "PUT" | "PATCH" | "DELETE") {
                return (
                    PermissionLevel::ConfirmedExecution,
                    format!("http_request with mutating method {method}"),
                    None,
                );
            }
        }

        if tool_name == "memory_manage" {
            let action = string_arg(request.arguments, "action")
                .unwrap_or("read")
                .to_lowercase();
            if action == "read" {
                return (PermissionLevel::ReadOnly, "memory read".into(), None);
            }
            return (
                PermissionLevel::ConfirmedExecution,
                format!("memory {action}"),
                None,
            );
        }

        if READ_ONLY_TOOLS.contains(tool_name) {
            return (PermissionLevel::ReadOnly, "read-only tool".into(), None);
        }
        if SAFE_ACTION_TOOLS.contains(tool_name) {
            return (PermissionLevel::SafeAction, "safe action tool".into(), None);
        }
        if CONFIRMED_TOOLS.contains(tool_name) {
            return (
                PermissionLevel::ConfirmedExecution,
                "tool requires confirmation".into(),
                None,
            );
        }

        if let Some(level) = classify_mcp_style_tool(tool_name) {
            return (
                level,
                format!("MCP-style {} tool", level.as_str().to_lowercase()),
                None,
            );
        }

        (
            PermissionLevel::SafeAction,
            "unclassified tool defaults to safe action".into(),
            None,
        )
    }

    fn audit(&self, request: &PermissionRequest<'_>, decision: &PermissionDecision) {
        if let Some(parent) = self.audit_log_path.parent() {
            if std::fs::create_dir_all(parent).is_err() {
                return;
            }
            set_owner_only_dir_permissions(parent);
        }

        let argument_keys = request
            .arguments
            .as_object()
            .map(|obj| {
                let mut keys = obj.keys().cloned().collect::<Vec<_>>();
                keys.sort();
                keys
            })
            .unwrap_or_default();

        let command_preview = request
            .command
            .map(ToOwned::to_owned)
            .unwrap_or_else(|| extract_command(request.arguments))
            .chars()
            .take(200)
            .collect::<String>();

        let record = json!({
            "timestamp": chrono::Utc::now().to_rfc3339(),
            "agent_id": request.agent_id.unwrap_or(""),
            "tool": request.tool_name,
            "action": decision.action.as_str(),
            "level": decision.level.as_str(),
            "reason": decision.reason,
            "matched_pattern": decision.matched_pattern,
            "dry_run": decision.dry_run,
            "argument_keys": argument_keys,
            "command_preview": command_preview,
            "request_metadata": {"source": request.source},
            "metadata": {
                "would_action": decision.would_action.as_ref().map(PermissionAction::as_str),
                "would_level": if decision.dry_run { Some(decision.level.as_str()) } else { None },
            },
        });

        if let Ok(mut file) = OpenOptions::new()
            .create(true)
            .append(true)
            .open(&self.audit_log_path)
        {
            let _ = writeln!(file, "{}", record);
            set_owner_only_permissions(&self.audit_log_path);
        }
    }
}

pub fn classify_shell_command(command: &str) -> (PermissionLevel, String, Option<String>) {
    let normalized = command.split_whitespace().collect::<Vec<_>>().join(" ");
    if normalized.is_empty() {
        return (
            PermissionLevel::SafeAction,
            "empty shell command".into(),
            None,
        );
    }

    for (label, pattern) in DANGEROUS_COMMAND_PATTERNS.iter() {
        if pattern.is_match(&normalized) {
            return (
                PermissionLevel::Dangerous,
                format!("dangerous shell command pattern: {label}"),
                Some((*label).into()),
            );
        }
    }

    (
        PermissionLevel::ConfirmedExecution,
        "shell command requires confirmation".into(),
        None,
    )
}

pub fn blocked_file_write_reason(arguments: &Value) -> Option<String> {
    let path = string_arg(arguments, "path")?;
    let path = Path::new(path);

    if has_path_traversal(path) {
        return Some("path traversal blocked".into());
    }
    if is_sensitive_path(path) {
        return Some("sensitive file write blocked".into());
    }
    if is_protected_path(path) {
        return Some("protected path write blocked".into());
    }

    None
}

fn action_for_level(level: PermissionLevel) -> PermissionAction {
    match level {
        PermissionLevel::Dangerous => PermissionAction::Deny,
        PermissionLevel::ConfirmedExecution => PermissionAction::RequireConfirmation,
        PermissionLevel::ReadOnly | PermissionLevel::SafeAction => PermissionAction::Allow,
    }
}

fn extract_command(arguments: &Value) -> String {
    for key in ["command", "cmd", "shell_command", "script"] {
        if let Some(value) = arguments.get(key) {
            return value
                .as_str()
                .map(ToOwned::to_owned)
                .unwrap_or_else(|| value.to_string());
        }
    }
    String::new()
}

fn string_arg<'a>(arguments: &'a Value, key: &str) -> Option<&'a str> {
    arguments.get(key).and_then(Value::as_str)
}

fn looks_like_shell_tool(tool_name: &str) -> bool {
    let normalized = tool_name.to_lowercase().replace('-', "_");
    ["shell", "terminal", "bash", "zsh", "run_command"]
        .iter()
        .any(|token| normalized.contains(token))
}

fn classify_mcp_style_tool(tool_name: &str) -> Option<PermissionLevel> {
    let normalized = tool_name.to_lowercase().replace('-', "_");
    if ["read_file", "get_file"]
        .iter()
        .any(|token| normalized.contains(token))
    {
        return Some(PermissionLevel::ReadOnly);
    }
    if [
        "write_file",
        "create_file",
        "edit_file",
        "patch_file",
        "delete_file",
        "remove_file",
        "apply_patch",
        "run_code",
        "execute_code",
        "python",
    ]
    .iter()
    .any(|token| normalized.contains(token))
    {
        return Some(PermissionLevel::ConfirmedExecution);
    }
    None
}

fn default_audit_log_path() -> PathBuf {
    std::env::var_os("OPENJARVIS_PERMISSION_AUDIT_LOG")
        .map(PathBuf::from)
        .or_else(|| {
            std::env::var_os("HOME")
                .map(PathBuf::from)
                .map(|home| home.join(".openjarvis/logs/permissions.log"))
        })
        .unwrap_or_else(|| PathBuf::from(".openjarvis/logs/permissions.log"))
}

#[cfg(unix)]
fn set_owner_only_permissions(path: &Path) {
    use std::os::unix::fs::PermissionsExt;

    if let Ok(metadata) = std::fs::metadata(path) {
        let mut permissions = metadata.permissions();
        permissions.set_mode(0o600);
        let _ = std::fs::set_permissions(path, permissions);
    }
}

#[cfg(not(unix))]
fn set_owner_only_permissions(_path: &Path) {}

#[cfg(unix)]
fn set_owner_only_dir_permissions(path: &Path) {
    use std::os::unix::fs::PermissionsExt;

    if let Ok(metadata) = std::fs::metadata(path) {
        let mut permissions = metadata.permissions();
        permissions.set_mode(0o700);
        let _ = std::fs::set_permissions(path, permissions);
    }
}

#[cfg(not(unix))]
fn set_owner_only_dir_permissions(_path: &Path) {}

static READ_ONLY_TOOLS: Lazy<HashSet<&'static str>> = Lazy::new(|| {
    HashSet::from([
        "browser_axtree",
        "browser_extract",
        "browser_screenshot",
        "calculator",
        "channel_list",
        "channel_status",
        "db_query",
        "file_read",
        "git_diff",
        "git_log",
        "git_status",
        "kg_neighbors",
        "kg_query",
        "knowledge_search",
        "list_scheduled_tasks",
        "retrieval",
        "scan_chunks",
        "think",
        "web_search",
    ])
});

static SAFE_ACTION_TOOLS: Lazy<HashSet<&'static str>> = Lazy::new(|| {
    HashSet::from([
        "agent_list",
        "browser_click",
        "browser_navigate",
        "browser_type",
        "digest_collect",
        "http_request",
        "knowledge_sql",
        "llm_tool",
        "mcp_adapter",
        "pdf_tool",
        "repl",
        "schedule_task",
        "skill_manage",
        "text_to_speech",
        "user_profile_manage",
    ])
});

static CONFIRMED_TOOLS: Lazy<HashSet<&'static str>> = Lazy::new(|| {
    HashSet::from([
        "agent_kill",
        "agent_send",
        "agent_spawn",
        "apply_patch",
        "cancel_scheduled_task",
        "channel_send",
        "code_interpreter",
        "code_interpreter_docker",
        "file_write",
        "git_commit",
        "kg_add_entity",
        "kg_add_relation",
        "memory_manage",
        "pause_scheduled_task",
        "repl",
        "resume_scheduled_task",
        "shell_exec",
        "storage_delete",
        "storage_put",
    ])
});

static DANGEROUS_COMMAND_PATTERNS: Lazy<Vec<(&'static str, Regex)>> = Lazy::new(|| {
    vec![
        (
            "recursive-root-delete",
            Regex::new(r"(?i)\brm\s+[^;&|]*-[^\s;&|]*r[f]?[^\s;&|]*\s+/(?:\s|$)").unwrap(),
        ),
        (
            "home-delete",
            Regex::new(r"(?i)\brm\s+[^;&|]*-[^\s;&|]*r[f]?[^\s;&|]*\s+~(?:/|\s|$)").unwrap(),
        ),
        (
            "force-clean-current-dir",
            Regex::new(r"(?i)\brm\s+[^;&|]*-[^\s;&|]*r[f]?[^\s;&|]*\s+\.(?:\s|$)").unwrap(),
        ),
        (
            "disk-format",
            Regex::new(r"(?i)\b(?:mkfs|fdisk|parted|diskutil\s+erase|diskutil\s+partition)\b")
                .unwrap(),
        ),
        (
            "raw-disk-write",
            Regex::new(r"(?i)\bdd\s+.*\bof=/dev/(?:disk|rdisk|sd|nvme)").unwrap(),
        ),
        (
            "privilege-escalation",
            Regex::new(r"(?i)\b(?:sudo|su)\b").unwrap(),
        ),
        (
            "permission-recursive-root",
            Regex::new(r"(?i)\b(?:chmod|chown|chgrp)\s+[^;&|]*-R[^;&|]*\s+/(?:\s|$)").unwrap(),
        ),
        (
            "shutdown",
            Regex::new(r"(?i)\b(?:shutdown|reboot|halt|poweroff)\b").unwrap(),
        ),
        (
            "fork-bomb",
            Regex::new(r":\s*\(\s*\)\s*\{\s*:\s*\|\s*:\s*&\s*\}\s*;?\s*:").unwrap(),
        ),
        (
            "pipe-to-shell",
            Regex::new(r"(?i)\b(?:curl|wget)\b.*\|\s*(?:sh|bash|zsh|python|perl)\b").unwrap(),
        ),
        (
            "git-destructive",
            Regex::new(r"(?i)\bgit\s+(?:reset\s+--hard|clean\s+-[^\s;&|]*f)").unwrap(),
        ),
        ("kill-all", Regex::new(r"(?i)\bkill\s+-9\s+-1\b").unwrap()),
    ]
});

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_shell_dangerous_detection() {
        let (level, reason, matched) = classify_shell_command("sudo rm -rf /");
        assert_eq!(level, PermissionLevel::Dangerous);
        assert!(reason.contains("dangerous shell command"));
        assert!(matched.is_some());
    }

    #[test]
    fn test_shell_requires_confirmation() {
        let (level, _, matched) = classify_shell_command("ls -la");
        assert_eq!(level, PermissionLevel::ConfirmedExecution);
        assert_eq!(matched, None);
    }

    #[test]
    fn test_file_write_protected_path() {
        let reason = blocked_file_write_reason(&json!({"path": "/etc/openjarvis.conf"}));
        assert_eq!(reason.as_deref(), Some("protected path write blocked"));
    }

    #[test]
    fn test_policy_writes_audit_log() {
        let dir = tempfile::tempdir().unwrap();
        let audit_path = dir.path().join("permissions.log");
        let policy = PermissionPolicy::new(&audit_path);
        let decision = policy.check(&PermissionRequest {
            tool_name: "calculator",
            arguments: &json!({"expression": "2+2"}),
            agent_id: Some("agent-1"),
            command: None,
            source: "test",
        });

        assert!(decision.allowed());
        let contents = std::fs::read_to_string(audit_path).unwrap();
        assert!(contents.contains("\"tool\":\"calculator\""));
        assert!(contents.contains("\"source\":\"test\""));
    }
}

File: src/openjarvis/security/permissions.py
"""Permission middleware for tool execution.

Phase 1 keeps this deliberately small: classify a tool request, make a
local allow/deny/confirmation decision, and write an append-only audit line.
"""

from __future__ import annotations

import json
import logging
import re
from dataclasses import dataclass, field
from datetime import datetime, timezone
from enum import IntEnum
from ipaddress import ip_address
from pathlib import Path
from typing import Any, Dict, Mapping, Optional
from urllib.parse import urlparse

from openjarvis.security.file_utils import secure_create

logger = logging.getLogger(__name__)


class PermissionLevel(IntEnum):
    """Coarse permission levels for tool requests."""

    READ_ONLY = 0
    SAFE_ACTION = 1
    CONFIRMED_EXECUTION = 2
    DANGEROUS = 3


@dataclass(slots=True)
class PermissionRequest:
    """Input to ``PermissionMiddleware.check``."""

    tool_name: str
    arguments: Mapping[str, Any] = field(default_factory=dict)
    agent_id: str = ""
    command: Optional[str] = None
    permission_ceiling: Optional[PermissionLevel] = None
    dry_run: bool = False
    metadata: Dict[str, Any] = field(default_factory=dict)


@dataclass(slots=True)
class PermissionDecision:
    """Decision returned by the permission middleware."""

    action: str
    level: PermissionLevel
    reason: str
    matched_pattern: Optional[str] = None
    dry_run: bool = False
    metadata: Dict[str, Any] = field(default_factory=dict)

    @property
    def allowed(self) -> bool:
        return self.action == "allow"

    @property
    def denied(self) -> bool:
        return self.action == "deny"

    @property
    def requires_confirmation(self) -> bool:
        return self.action == "require_confirmation"


class PermissionMiddleware:
    """Classify tool calls and return coarse permission decisions."""

    _READ_ONLY_TOOLS = frozenset(
        {
            "browser_axtree",
            "browser_extract",
            "browser_screenshot",
            "calculator",
            "channel_list",
            "channel_status",
            "db_query",
            "file_read",
            "git_diff",
            "git_log",
            "git_status",
            "kg_neighbors",
            "kg_query",
            "knowledge_search",
            "list_scheduled_tasks",
            "retrieval",
            "scan_chunks",
            "terminal_context",
            "terminal_error_summary",
            "terminal_history",
            "terminal_suggested_fixes",
            "think",
            "web_search",
        }
    )
    _SAFE_ACTION_TOOLS = frozenset(
        {
            "agent_list",
            "browser_click",
            "browser_navigate",
            "browser_type",
            "digest_collect",
            "http_request",
            "knowledge_sql",
            "llm_tool",
            "mcp_adapter",
            "pdf_tool",
            "repl",
            "schedule_task",
            "skill_manage",
            "text_to_speech",
            "user_profile_manage",
            "vision_screenshot_capture",
        }
    )
    _CONFIRMED_TOOLS = frozenset(
        {
            "agent_kill",
            "agent_send",
            "agent_spawn",
            "apply_patch",
            "cancel_scheduled_task",
            "channel_send",
            "code_interpreter",
            "code_interpreter_docker",
            "file_write",
            "git_commit",
            "kg_add_entity",
            "kg_add_relation",
            "memory_manage",
            "pause_scheduled_task",
            "repl",
            "resume_scheduled_task",
            "shell_exec",
            "storage_delete",
            "storage_put",
            "terminal_suggested_command_approval",
            "voice_global_hotkey_listener",
            "voice_microphone_capture",
        }
    )
    _DANGEROUS_TOOLS = frozenset()

    _DANGEROUS_COMMAND_PATTERNS = tuple(
        (label, re.compile(pattern, re.IGNORECASE))
        for label, pattern in (
            (
                "recursive-root-delete",
                r"\brm\s+[^;&|]*-[^\s;&|]*r[f]?[^\s;&|]*\s+/(?:\s|$)",
            ),
            ("home-delete", r"\brm\s+[^;&|]*-[^\s;&|]*r[f]?[^\s;&|]*\s+~(?:/|\s|$)"),
            (
                "force-clean-current-dir",
                r"\brm\s+[^;&|]*-[^\s;&|]*r[f]?[^\s;&|]*\s+\.(?:\s|$)",
            ),
            (
                "disk-format",
                r"\b(?:mkfs|fdisk|parted|diskutil\s+erase|diskutil\s+partition)\b",
            ),
            ("raw-disk-write", r"\bdd\s+.*\bof=/dev/(?:disk|rdisk|sd|nvme)"),
            ("privilege-escalation", r"\b(?:sudo|su)\b"),
            (
                "permission-recursive-root",
                r"\b(?:chmod|chown|chgrp)\s+[^;&|]*-R[^;&|]*\s+/(?:\s|$)",
            ),
            ("shutdown", r"\b(?:shutdown|reboot|halt|poweroff)\b"),
            ("fork-bomb", r":\s*\(\s*\)\s*\{\s*:\s*\|\s*:\s*&\s*\}\s*;?\s*:"),
            ("pipe-to-shell", r"\b(?:curl|wget)\b.*\|\s*(?:sh|bash|zsh|python|perl)\b"),
            ("git-destructive", r"\bgit\s+(?:reset\s+--hard|clean\s+-[^\s;&|]*f)"),
            ("kill-all", r"\bkill\s+-9\s+-1\b"),
        )
    )

    def __init__(
        self,
        *,
        audit_log_path: Optional[Path | str] = None,
        dry_run: bool = False,
        mode_registry: Any = None,
    ) -> None:
        self._audit_log_path = (
            Path(audit_log_path).expanduser()
            if audit_log_path is not None
            else Path.home() / ".openjarvis" / "logs" / "permissions.log"
        )
        self._dry_run = dry_run
        self._mode_registry = mode_registry

    @property
    def audit_log_path(self) -> Path:
        return self._audit_log_path

    def check(self, request: PermissionRequest) -> PermissionDecision:
        """Return the permission decision for *request* and audit it."""
        level, reason, matched_pattern = self.classify(request)
        ceiling = request.permission_ceiling
        if ceiling is not None and level > ceiling:
            reason = (
                f"permission level {level.name} exceeds agent ceiling "
                f"{ceiling.name}: {reason}"
            )
            level = PermissionLevel.DANGEROUS
            matched_pattern = matched_pattern or "agent-permission-ceiling"
        action = self._action_for_level(level)
        dry_run = self._dry_run or request.dry_run

        metadata: Dict[str, Any] = {}
        if dry_run:
            metadata["would_action"] = action
            metadata["would_level"] = level.name
            action = "allow"
            reason = f"dry_run: would {metadata['would_action']} ({reason})"

        decision = PermissionDecision(
            action=action,
            level=level,
            reason=reason,
            matched_pattern=matched_pattern,
            dry_run=dry_run,
            metadata=metadata,
        )
        self._audit(request, decision)
        return decision

    def classify(
        self,
        request: PermissionRequest,
    ) -> tuple[PermissionLevel, str, Optional[str]]:
        """Classify a permission request without applying dry-run behavior."""
        tool_name = request.tool_name
        arguments = request.arguments if isinstance(request.arguments, Mapping) else {}
        active_mode = self._active_mode(request)

        privacy_block = self._privacy_mode_block(tool_name, arguments, active_mode)
        if privacy_block is not None:
            reason, matched_pattern = privacy_block
            return PermissionLevel.DANGEROUS, reason, matched_pattern

        command = self._extract_command(request, arguments)
        if tool_name == "shell_exec" or self._looks_like_shell_tool(tool_name):
            level, reason, matched_pattern = self.classify_shell_command(command)
            if (
                tool_name != "shell_exec"
                and level == PermissionLevel.CONFIRMED_EXECUTION
            ):
                reason = f"MCP shell-style tool requires confirmation: {tool_name}"
            return level, reason, matched_pattern

        if tool_name == "http_request":
            method = str(arguments.get("method", "GET")).upper()
            if method in {"POST", "PUT", "PATCH", "DELETE"}:
                return (
                    PermissionLevel.CONFIRMED_EXECUTION,
                    f"http_request with mutating method {method}",
                    None,
                )

        if tool_name == "memory_manage":
            action = str(arguments.get("action", "read")).lower()
            if action == "read":
                return PermissionLevel.READ_ONLY, "memory read", None
            return PermissionLevel.CONFIRMED_EXECUTION, f"memory {action}", None

        if tool_name in self._READ_ONLY_TOOLS:
            return PermissionLevel.READ_ONLY, "read-only tool", None
        if tool_name in self._SAFE_ACTION_TOOLS:
            return PermissionLevel.SAFE_ACTION, "safe action tool", None
        if tool_name in self._CONFIRMED_TOOLS:
            return (
                PermissionLevel.CONFIRMED_EXECUTION,
                "tool requires confirmation",
                None,
            )
        if tool_name in self._DANGEROUS_TOOLS:
            return PermissionLevel.DANGEROUS, "dangerous tool", None

        mcp_level = self._classify_mcp_style_tool(tool_name)
        if mcp_level is not None:
            return mcp_level, f"MCP-style {mcp_level.name.lower()} tool", None

        return (
            PermissionLevel.SAFE_ACTION,
            "unclassified tool defaults to safe action",
            None,
        )

    def _active_mode(self, request: PermissionRequest) -> Any:
        mode = request.metadata.get("active_mode")
        if mode is not None and not isinstance(mode, str):
            return mode
        registry = self._mode_registry
        if registry is None:
            return None
        try:
            mode_id = request.metadata.get("active_mode_id")
            if mode_id:
                return registry.get_mode(str(mode_id))
            return registry.get_active_mode().mode
        except Exception:
            return None

    def _privacy_mode_block(
        self,
        tool_name: str,
        arguments: Mapping[str, Any],
        active_mode: Any,
    ) -> Optional[tuple[str, str]]:
        if getattr(active_mode, "id", "") != "privacy":
            return None

        normalized = tool_name.lower().replace("-", "_")
        if "mcp" in normalized:
            return "privacy mode blocks remote MCP tool execution", "privacy-remote-mcp"

        if tool_name == "http_request":
            url = str(arguments.get("url", ""))
            if not self._is_localhost_url(url):
                return (
                    "privacy mode blocks outbound non-localhost HTTP requests",
                    "privacy-network-localhost-only",
                )

        if normalized in {"web_search", "browser_navigate"}:
            url = str(arguments.get("url", ""))
            if not url or not self._is_localhost_url(url):
                return (
                    "privacy mode blocks outbound non-localhost networking",
                    "privacy-network-localhost-only",
                )

        if normalized in {"vision_screenshot_capture", "browser_screenshot"}:
            return (
                "privacy mode requires approval before screenshot capture",
                "privacy-vision-screenshot",
            )

        if normalized == "voice_microphone_capture" and not arguments.get(
            "explicit_approval"
        ):
            return (
                "privacy mode requires explicit approval before voice capture",
                "privacy-voice-capture",
            )

        if normalized == "voice_global_hotkey_listener":
            return (
                "privacy mode disables the global voice hotkey listener",
                "privacy-global-voice-hotkey",
            )

        return None

    @staticmethod
    def _is_localhost_url(url: str) -> bool:
        try:
            parsed = urlparse(url)
        except Exception:
            return False
        host = parsed.hostname
        if not host:
            return False
        normalized = host.strip("[]").lower()
        if normalized in {"localhost", "localhost.localdomain"}:
            return True
        try:
            parsed_ip = ip_address(normalized)
        except ValueError:
            return False
        return parsed_ip.is_loopback

    def classify_shell_command(
        self,
        command: str,
    ) -> tuple[PermissionLevel, str, Optional[str]]:
        """Classify a shell command using known dangerous patterns."""
        normalized = " ".join(command.strip().split())
        if not normalized:
            return PermissionLevel.SAFE_ACTION, "empty shell command", None

        for label, pattern in self._DANGEROUS_COMMAND_PATTERNS:
            if pattern.search(normalized):
                return (
                    PermissionLevel.DANGEROUS,
                    f"dangerous shell command pattern: {label}",
                    label,
                )

        return (
            PermissionLevel.CONFIRMED_EXECUTION,
            "shell command requires confirmation",
            None,
        )

    @staticmethod
    def _action_for_level(level: PermissionLevel) -> str:
        if level == PermissionLevel.DANGEROUS:
            return "deny"
        if level == PermissionLevel.CONFIRMED_EXECUTION:
            return "require_confirmation"
        return "allow"

    @staticmethod
    def _extract_command(
        request: PermissionRequest,
        arguments: Mapping[str, Any],
    ) -> str:
        if request.command is not None:
            return request.command
        for key in ("command", "cmd", "shell_command", "script"):
            raw = arguments.get(key)
            if raw is not None:
                return raw if isinstance(raw, str) else str(raw)
        return ""

    @staticmethod
    def _looks_like_shell_tool(tool_name: str) -> bool:
        normalized = tool_name.lower().replace("-", "_")
        shell_tokens = ("shell", "bash", "zsh", "run_command")
        return any(token in normalized for token in shell_tokens)

    @staticmethod
    def _classify_mcp_style_tool(tool_name: str) -> Optional[PermissionLevel]:
        normalized = tool_name.lower().replace("-", "_")
        if any(token in normalized for token in ("read_file", "get_file")):
            return PermissionLevel.READ_ONLY
        if any(
            token in normalized
            for token in (
                "write_file",
                "create_file",
                "edit_file",
                "patch_file",
                "delete_file",
                "remove_file",
                "apply_patch",
                "run_code",
                "execute_code",
                "python",
            )
        ):
            return PermissionLevel.CONFIRMED_EXECUTION
        return None

    def _audit(
        self,
        request: PermissionRequest,
        decision: PermissionDecision,
    ) -> None:
        try:
            path = secure_create(self._audit_log_path)
            record = {
                "timestamp": datetime.now(timezone.utc).isoformat(),
                "agent_id": request.agent_id,
                "tool": request.tool_name,
                "action": decision.action,
                "level": decision.level.name,
                "reason": decision.reason,
                "matched_pattern": decision.matched_pattern,
                "dry_run": decision.dry_run,
                "argument_keys": sorted(
                    str(key)
                    for key in (
                        request.arguments.keys()
                        if isinstance(request.arguments, Mapping)
                        else []
                    )
                ),
                "command_preview": self._command_preview(request),
                "request_metadata": request.metadata,
                "metadata": decision.metadata,
            }
            with path.open("a", encoding="utf-8") as fh:
                fh.write(json.dumps(record, sort_keys=True) + "\n")
        except Exception:
            logger.debug("Failed to write permission audit log", exc_info=True)

    @staticmethod
    def _command_preview(request: PermissionRequest) -> str:
        command = request.command
        if command is None and isinstance(request.arguments, Mapping):
            raw = request.arguments.get("command", "")
            command = raw if isinstance(raw, str) else ""
        return (command or "")[:200]


__all__ = [
    "PermissionDecision",
    "PermissionLevel",
    "PermissionMiddleware",
    "PermissionRequest",
]
